### ISSUE TITLE
Question on Multimodal attention mask

### ISSUE THREAD

**Issue by ipheiman** [jefferyYu/TomBERT#2]
Hi authors, thank you for the well-written paper and  detailed documentation of your work. I have a question regarding the multimodal attention mask.

**Under TomBERT/my_bert/mm_modelling.py:** 

```
class MBertForMMSequenceClassification(PreTrainedBertModel):
    """BERT model for classification with text and image inputs, pooling-1+text (MBERT I)
    """

```
    def __init__(self, config, num_labels=2, pooling="cls"):
        super(MBertForMMSequenceClassification, self).__init__(config)
        self.num_labels = num_labels
        self.pooling = pooling
        self.bert = BertModel(config)
        self.dropout = nn.Dropout(config.hidden_dropout_prob)
        self.vismap2text = nn.Linear(2048, config.hidden_size)
        #self.img_attention = BertLayer(config)
        self.comb_attention = MultimodalEncoder(config)
        if pooling == "cls":
            self.text_pooler = BertText1Pooler(config)
            self.classifier = nn.Linear(config.hidden_size, num_labels)
        elif pooling == "first":
            self.img_pooler = BertPooler(config)
            self.classifier = nn.Linear(config.hidden_size, num_labels)
        else:
            self.text_pooler = BertText1Pooler(config)
            self.img_pooler = BertPooler(config)
            self.classifier = nn.Linear(config.hidden_size * 2, num_labels)

        self.apply(self.init_bert_weights)

    def forward(self, input_ids, s2_input_ids, visual_embeds_att, token_type_ids=None, s2_type_ids=None,
                attention_mask=None, s2_mask=None, added_attention_mask=None, labels=None, copy_flag=False):
        # Concatenate Bert-based Text, Text-Aware Image and Image-Aware Text
        sequence_output, pooled_output = self.bert(input_ids, token_type_ids, attention_mask,
                                                   output_all_encoded_layers=False)

        # apply entity-based attention mechanism to obtain different image representations
        vis_embed_map = visual_embeds_att.view(-1, 2048, 49).permute(0, 2, 1)  # self.batch_size, 49, 2048
        vis_pooled_output, _ = vis_embed_map.max(1)  # self.batch_size, 2048
        converted_vis_embed_map = self.vismap2text(vis_pooled_output)  # self.batch_size, hidden_dim

        transpose_img_embed = converted_vis_embed_map.unsqueeze(1)
        text_img_output = torch.cat((transpose_img_embed, sequence_output), dim=1)

        comb_attention_mask = added_attention_mask[:, 48:]  # only the first dimension is for image
        extended_attention_mask = comb_attention_mask.unsqueeze(1).unsqueeze(2)
        extended_attention_mask = extended_attention_mask.to(dtype=next(self.parameters()).dtype)  # fp16 compatibility
        extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
`
I am trying to understand the extended_attention_mask here.

- So far, I understand that **added_attention_mask** refers to [1]*(len(input_ids)  + 49) followed by zero-pad to get length 113. (64 values for input_id mask and 49 values for image patches)
- And for each input, this would vary since the input_ids will have varying lengths. So why does comb_attention_mask specifically start from index 48?
- And why does extended_attention_mask need to multiply -10000.0

Hope to hear from you, thanks so much!

**Comment by jefferyYu** [comment]
Hi there,

Thanks for your interest in our work!

1. For added_attention_mask, the first 49 mask variables refer to the 49 image regions (all of them are 1), and the following mask variables refer to the text mask variables, including valid text inputs followed by zero-padding.
2.  For comb_attention_mask in the mBERT model, the first mask variable refers to the image representation (note that this is the pooled image representation), and the next mask variables refer to the text mask variables. Hence, it starts from 48, because the first mask variable is for the image, and the remaining mask variables are for the text.
3.  For multiplying -10000.0, this is a common practice in the implementation of Transformer-based models, and you may have a better understanding after carefully reading the Transformer repos in HuggingFace. Since the mask variables for the padding positions are 0, the result of (1.0 - extended_attention_mask)* -10000.0 will be -10000.0 and then applying a softmax over extended_attention_mask will make the value of the padding positions to be close to 0.

**Comment by ipheiman** [comment]
Hi author,
I understand much clearer now thanks for your detailed explanation!
1. The multimodal feature is concatenated in the order of (img , text) and hence the first 49 variables will always be 1 in added_attention_mask. (Previously I thought it was (text,img))
2. After the pooled_output, the 49 image patches are consolidated to 1 "patch". so the combined attention mask from 48 to 49 is for that 1 patch, and 49 to 113 is for text.
3. This is cool! I just read about it here [https://github.com/huggingface/transformers/issues/542](url) . So 0s and 1s as mask variables is not good enough because when softmax is applied, e^0 is almost as huge as e^1. Hence 0s are converted to very negative values of -10000 such that it will give a very low weightage after softmax.

Thanks again for the prompt reply! :)
